File: 2023/python/day02a/main.py
import os

d = {
    "red": 12,
    "green": 13,
    "blue": 14,
}

f = open(os.path.join(os.path.dirname(__file__), 'input.txt'), "r")

input_data = f.read().splitlines()
sum = 0
for line in input_data:
    game, rounds = line.split(": ")
    id = int(game.split(' ')[1])

    rounds = rounds.split("; ")
    cont = False
    for round in rounds:
        cubes = round.split(", ")
        for cube in cubes:
            cube = cube.split(' ')
            color = cube[1]
            number = int(cube[0])
            if d[color] < number:
                cont = True
                break
    if cont:
        continue
    else:
        sum += id

print(sum)

f.close()

File: 2023/rust/day02a/src/main.rs
fn main() {
    let input: Vec<_> = include_str!("input.txt")
        .split('\n')
        .collect();
    let mut sum: i32 = 0;

    for line in input {
        let game: Vec<&str> = line.split(": ").collect();
        let game_id: &str = game[0].split(' ').collect::<Vec<_>>()[1];
        let id: i32 = game_id.parse::<i32>().unwrap();

        let records: Vec<&str> = game[1].split("; ").collect();
        let mut add: bool = true;
        for record in records {
            let mut red: i32 = 12;
            let mut green: i32 = 13;
            let mut blue: i32 = 14;

            let cubes: Vec<&str> = record.split(", ").collect();
            for cube in cubes {
                let components: Vec<&str> = cube.split(' ').collect();
                let color: &str = components[1];
                let number: i32 = components[0].parse::<i32>().unwrap();
                match color {
                    "green" => {
                        green -= number;
                    },
                    "red" => {
                        red -= number;
                    },
                    "blue" => {
                        blue -= number;
                    }
                    _ => panic!("Invalid color")
                }
            }

            add = if red < 0 || blue < 0 || green < 0 { false } else { add };
        }

        sum = if add { sum + id } else { sum };
    }
    println!("{}", sum);
}